### PR TITLE
MONGOID-4525 Fix ability to reset Hash field to empty Hash (on master)

### PR DESCRIPTION
**Pull request by mdehoog**
Combination of #4487 / #4510: forward-port to master.

### CODE REVIEW

**Comment by p-mongo** [comment]
Thank you for this.

**Comment by mdehoog** [comment]
@johnnyshields I don't believe we do. The original bug fix was to allow setting a value to an empty hash; the root cause was due to the special handling of hashes such that keys are merged. This special handling doesn't apply to arrays, and setting empty arrays should work fine.

**Comment by p-mongo** [comment]
Adding additional test cases, one problem with this implementation is it merges leaf hash values instead of writing them as is (which is what mongo shell does).

I.e. if I have `{'a' => {'b' => 'c'}` and I set `'a' => {'d' => 'e'}`, the result should be `{'a' => {'d' => 'e'}}` but it currently is `{'a' => {'b' => 'c', 'd' => 'e'}}`. (To achieve the merge on the leaf level the operation should be to set `'a.d' => 'e'`.)

**Comment by mdehoog** [comment]
@p-mongo your new tests fail on master also. Perhaps it would be better to track this bug in a separate PR and issue, rather than conflating two fixes into one?

**Comment by p-mongo** [comment]
Should be fixed now. Do you want to make that ticket?

File: lib/mongoid/persistable/settable.rb [hunk 1]
```diff
@@ -23,17 +51,39 @@ def set(setters)
...
-              process_attribute(field.to_s, value)
-            else
-              process_attribute(field.to_s, field_and_value_hash[field])
+              # now we are on the leaf level, perform the set
+              # and overwrite whatever was on this level before
+              cur_value[field_seq.shift] = value
+
+              # and set value to the value of the top level field
+              # because this is what we pass to $set
+              value = old_value
```
Comment (mdehoog) [inline lib/mongoid/persistable/settable.rb]: now that the bulk of this `set` method consists of logic handling nested hashing, perhaps section should be pulled out into a separate method specifically for merging hashes?

**Comment by mdehoog** [comment]
I went ahead and opened an issue here: https://jira.mongodb.org/browse/MONGOID-4593.

Closing in favor of #4518, which contains the same commits and is specifically a fix for the above issue.

**Pull request closed without merging**